### ISSUE TITLE
[rush] Add an interactive console for invoking project commands, eg. watch scripts and dev-servers

### ISSUE THREAD

**Issue by wbern** [microsoft/rushstack#2265]
**Is this a feature or a bug?**

- [X] Feature
- [ ] Bug

**Please describe the actual behavior.**

Right now, it is a little difficult to quickly run a couple of watch scripts on packages inside a Rush monorepo. Lerna does this fairly well with `lerna run`, but maybe there's a better way yet to be discovered.

**What is the expected behavior?**

I would like to be able to run watch scripts in several packages in a convenient manner. I made a tool which does this _pretty_ well, available as a prerelease at the time of writing, installable via `npm i -g @telia/rush-select@next` as version 0.2.0.

![image](https://user-images.githubusercontent.com/25477382/98422094-192f3800-208b-11eb-9a63-c77c94f44f59.png)

* Use the right/left arrow keys to move which script you want to execute for each package.
* The "smart" mode in the rush build option will perform a `rush build` command, but only for each package (using `--to`) that you have selected to do something in the prompt, including the dependencies they have. This is immensely useful at our org where we have a Rush repo of 114 modules at the time of writing (build time 3 mins total for a non-production build)
* The submitted results are saved per the path of where you store your rush repo on your machine, so it is easy to stop and start it again with what you entered last time.
* When you select a script for a project, the project will be duplicated on a new line, allowing you to choose multiple scripts in the same project if that fits your use-case.

Source code can be found here: https://github.com/TeliaSweden/rush-select/tree/new-prompt-type

I would like to see if anyone has any interest in bringing a prompt like this into Rush, and how it could look like.

Please try out my prompt to see if you have any feedback on how it could be improved. I have based it on the Enquirer [scale prompt](https://github.com/enquirer/enquirer#scale-prompt), but it has been extensively rewritten to work for our purpose.

All feedback and thoughts are welcome!

**Comment by iclanton** [comment]
This looks pretty awesome. I'll try it out today.

**Comment by wbern** [comment]
@iclanton released 0.3.0 that addresses process hanging and some tiny other nitpicks.

**Comment by wbern** [comment]
I released 0.4.0 under `next` which fixes some vertical scrolling issues.

**Comment by wbern** [comment]
Released 0.5.0 under `next` (install via `npm i -g @telia/rush-select@next`

I have cleaned up a few things, started migration to TypeScript to maybe fit into the rushjs project better in the future, and added more visual hints and not just colors to the prompt. Also fixed a save/load issue.

![image](https://user-images.githubusercontent.com/25477382/97760516-9abf1d00-1b03-11eb-84ec-00ee7789bb2e.png)

**Comment by wbern** [comment]
Released 0.6.0. As usual, install through `npm i -g @telia/rush-select@next`

* Fixed issues selecting scripts while filtering
* Fixed some visual issues while filtering
* Reached almost completed state of migration to TypeScript.
* Corrections to some TypeScript interfaces

Looking pretty good now!

![rush-select-mock-projects](https://user-images.githubusercontent.com/25477382/98422280-8c38ae80-208b-11eb-8ce8-5b33c482093f.gif)

**Comment by wbern** [comment]
FYI, this package is now generally available via plain `npm install -g @telia/rush-select`.

**Comment by wbern** [comment]
I was inspired to do something about the additional rows that were being created to execute multiple scripts, because most users don't need it and were confused by it. Someone even thought it was a bug. 😄 

I started working on an alternate way to display the selection. For now, it does not mess with the expected behavior that users are accustomed to, but if you use the space bar, you'll be able to multi-select things.

See the gif. It runs scripts properly, but the save/load mechanism is still not quite there (you'll see some remnants of the old way things worked, although it should still _work_). If you want to try it early, do `npm install -g @telia/rush-select@next`.

![rush-select-multi-select](https://user-images.githubusercontent.com/25477382/101995180-8bb5b800-3cc8-11eb-9b44-dfb42d5229e9.gif)

**Comment by wbern** [comment]
@iclanton @octogonz I think we need to come up with a decision of whether this should be worked into rushjs or if we should stay on the contribution track.

The benefits of me staying outside rush is that I could manage the code base more freely, and I could also later choose to work with monorepos that aren't necessarily using rush. Theoretically, you guys could also still integrate the prompt by installing rush-select as a dependency to the cli and running it from rush through that way (so people won't have to download the extra tool).

Or we integrate it harder into rush, but I'm not sure about how philosophy of rush compares to a semi-hacky and flexible tool like this, plus you have bigger designs ongoing for build orchestration.

**Comment by mikecann** [comment]
I am very new to rush but gotta throw my 2 cents in here and say this is an awesome tool and probably should be more widely known for others coming to rush, either being integrated or at the very least a page mentioning it in the docs 👍

**Comment by octogonz** [comment]
> Or we integrate it harder into rush, but I'm not sure about how philosophy of rush compares to a semi-hacky and flexible tool like this, plus you have bigger designs ongoing for build orchestration.

@wbern I'm in favor of officially incorporating your project, if you want to go that way. We were just waiting to see some other people from the community using it and saying they found it to be valuable.  

The tool itself seems rather scoped, so I'm not too concerned about maintenance cost as other Rush features evolve.

If we integrate it, we probably would want to do some cleanup:
- Update the files to follow the `@rushstack/eslint-config` coding conventions
- Replace some utilities with Rush equivalents to avoid increasing the installation footprint too much (e.g. maybe Yargs could be replaced by `ts-command-line`, etc)
- Write some docs

Another idea would be to model it as its own NPM package (e.g. `@rushstack/rush-select`), but officially maintained in the Rush repo.  Might be interesting to compare the pros/cons of that as well.

**Comment by wbern** [comment]
@octogonz I wrote this for rushjs, and I'd love to contribute to rush itself with it.

After getting my latest changes out the door, I will be focusing on the pointers you gave me. Excited to get this into rush!

A good action to take by someone else could be to open a contrib page on the rushjs site in the mean time, rush-select will likely not be the single thing to put on there based on what I've observed.
